src/main/scala/chisel3/util/BitSet.scala

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.util
+
+trait BitSet { bsf =>
+  val terms: Set[BitPat]
+
+  /**
+    * Get specified width of said BitSet
+    *
+    * All BitPat contained in one set should have the same width
+    */
+  def getWidth: Int = {
+    assert(terms.map(_.width).size <= 1)
+    // set width = 0 if terms is empty.
+    terms.headOption.map(_.width).getOrElse(0)
+  }
+
+  override def toString: String = terms.toSeq.sortBy((t: BitPat) => (t.mask, t.value)).mkString("\n")
+
+  /**
+    * @return whether the BitSetFamily is empty (i.e. no value matches)
+    */
+  def isEmpty: Boolean = terms.forall(_.isEmpty)
+
+  /**
+    * @return whether this BitSetFamily overlap with that BitSetFamily, i.e. !(intersect.isEmpty)
+    */
+  def overlap(that: BitSet): Boolean = {
+    !bsf.terms.flatMap(a => that.terms.map(b => (a, b))).forall { case (a, b) => !a.overlap(b) }
+  }
+
+  /**
+    * @param that BitSetFamily to be covered
+    * @return whether this BitSetFamily covers that (i.e. forall b matches that, b also matches this)
+    */
+  def cover(that: BitSet): Boolean =
+    that.subtract(this).isEmpty
+
+  /**
+    * @return a BitSetFamily that only match a value when both operand match
+    */
+  def intersect(that: BitSet): BitSet =
+    bsf.terms
+      .flatMap(a => that.terms.map(b => a.intersect(b)))
+      .filterNot(_.isEmpty)
+      .fold(BitSet.emptyBitSet)(_.union(_))
+
+  /**
+    * Subtract that from this BitSetFamily
+    * @param that subtrahend
+    */
+  def subtract(that: BitSet): BitSet =
+    bsf.terms.map { a =>
+      that.terms.map(b => a.subtract(b)).fold(a)(_.intersect(_))
+    }.filterNot(_.isEmpty).fold(BitSet.emptyBitSet)(_.union(_))
+
+  /**
+    * Union of two BitSetFamily
+    */
+  def union(that: BitSet): BitSet = new BitSet {
+    val terms = bsf.terms ++ that.terms
+  }
+
+  /**
+    * Test whether two BitSetFamily matches the same set of value
+    *
+    * Caution: This method can be very expensive compared to ordinary == operator between two Objects
+    */
+  override def equals(obj: Any): Boolean = {
+    obj match {
+      case that: BitSet => if (this.getWidth == that.getWidth) this.cover(that) && that.cover(this) else false
```

Comment:
```suggestion
      case that: BitSet => this.getWidth == that.getWidth && this.cover(that) && that.cover(this)
```
Scala short circuits boolean operators so this is exactly equivalent and IMO a little more clear.

Reply:
My bad!